### ISSUE TITLE
report helpful message when newuidmap is not privileged

### ISSUE THREAD

**Issue by specing** [lxc/lxc#296]
After a system update (shadow and LXC weren't included), LXC refuses to start previously-working virtual machines.

It seems a setuid is missed somewhere in lxc-usernsexec, but the same build worked before the system update.

Here is the same problem on a grsecurity machine (includes exec log and some additional printfs in newuidmap and LXC):  https://forums.grsecurity.net/viewtopic.php?f=3&t=4026&p=14320#p14320

distro: Gentoo
kernel: aufs-sources-3.13.6 (same that worked before the update)
LXC versions tested: lxc-1.0.0 tag from git (2014.02.20), latest git, 2014.06.06 git

on the grsecurity machine, kernels gentoo-sources-3.14.14 and hardened-sources-3.14.5-r2 have been tested, neither works.

I'm fairly convinced the problem is not in the kernel or LXC, but "somewhere in between". I have tried downgrading both glibc and pam to working versions, but it still doesen't work.

From the previously-working machine:

lxc-start -l DEBUG -o /tmp/lxctest.log -n usernstest -f usernstest.conf

```
  newuidmap: Target 25276 is owned by a different user
  Error mapping uids
  error mapping child
  setgid: Invalid argument
  lxc-start: Failed to chown /dev/pts/55
  lxc-start: Failed to shift tty into container
  lxc-start: failed to initialize the container
```

(I have to reset the terminal afterwards, as input is not being echoed for some reason -- missing cleanup?)

```
  lxc-start 1408279900.058 INFO     lxc_utils - XDG_RUNTIME_DIR isn't set in the environment.
  lxc-start 1408279900.058 WARN     lxc_log - lxc_log_init called with log already initialized
  lxc-start 1408279900.058 INFO     lxc_confile - read uid map: type u nsid 0 hostid 262144 range 65536
  lxc-start 1408279900.058 INFO     lxc_confile - read uid map: type g nsid 0 hostid 262144 range 65536
  lxc-start 1408279900.058 INFO     lxc_utils - XDG_RUNTIME_DIR isn't set in the environment.
  lxc-start 1408279900.059 DEBUG    lxc_conf - allocated pty '/dev/pts/55' (5/6)
  lxc-start 1408279900.059 INFO     lxc_conf - tty's configured
  lxc-start 1408279900.059 DEBUG    lxc_start - sigchild handler set
  lxc-start 1408279900.059 DEBUG    lxc_console - opening /dev/tty for console peer
  lxc-start 1408279900.059 INFO     lxc_caps - Last supported cap was 34
  lxc-start 1408279900.059 DEBUG    lxc_console - using '/dev/tty' as console
  lxc-start 1408279900.059 DEBUG    lxc_console - 25274 got SIGWINCH fd 11
  lxc-start 1408279900.059 DEBUG    lxc_console - set winsz dstfd:8 cols:177 rows:61
  lxc-start 1408279900.100 ERROR    lxc_conf - Failed to chown /dev/pts/55
  lxc-start 1408279900.100 ERROR    lxc_start - Failed to shift tty into container
  lxc-start 1408279900.100 INFO     lxc_utils - XDG_RUNTIME_DIR isn't set in the environment.
  lxc-start 1408279900.100 ERROR    lxc_start - failed to initialize the container
```

**Comment by specing** [comment]
working machine:        crw--w---- 1 root tty 136, 8 Nov 23 01:23 /dev/pts/8
non-working machine: crw--w---- 1 root tty 136, 18 Nov 22 23:32 /dev/pts/18

**Comment by hallyn** [comment]
If this is still an issue, (since you mention rebuilding shadow anyway :) I'd recomend adding extra debug info in newuidmap() to show the getuid() and pw->pw_uid and pw->pw_name.

I assume these are user-namespaced containers owned by the root user?

**Comment by specing** [comment]
@hallyn 4.0.8 kernel+grsec kernel does not work (even when everything disabled via sysctl), but the 4.0.9 (without grsec) kernel does work (same config sans grsec and a few new features)

I'll add the prints when I reboot back

**Comment by specing** [comment]
--- newuidmap.c 2016-04-10 00:15:17.676630496 +0200
+++ newuidmap.new.c     2016-04-10 00:26:07.583330173 +0200
@@ -160,6 +160,14 @@
            (getgid() != pw->pw_gid) ||
            (pw->pw_uid != st.st_uid) ||
            (pw->pw_gid != st.st_gid)) {
+
+               printf ("getuid(): %u\n", getuid());
+               printf ("pw->pw_uid: %u\n", pw->pw_uid);
+               printf ("pw->pw_gid: %u\n", pw->pw_gid);
+               printf ("pw->pw_name: %s\n", pw->pw_name);
+               printf ("st.st_uid: %d\n", st.st_uid);
+               printf ("st.st_gid: %d\n", st.st_gid);
+
                fprintf(stderr, _( "%s: Target %u is owned by a different user\n" ),
                        Prog, target);
                return EXIT_FAILURE;

**Comment by specing** [comment]
getuid(): 0
pw->pw_uid: 0
pw->pw_gid: 0
pw->pw_name: root
st.st_uid: 0
st.st_gid: 65500
newuidmap: Target 19325 is owned by a different user

Interestingly, 65500 is the group that has permission to view /proc (grsecurity)

**Comment by specing** [comment]
Apparently it is shadow-4.2.1 that is not working, just "downgraded" to shadow-4.1.5.1-r1 and it works.

**Comment by hallyn** [comment]
Could you show a diff of newuidmap.c, newgidmap.c, and lib/commonio.c between those versions?

**Comment by specing** [comment]
newuidmap.c does not exist in 4.1.5.1-r1 (grep -iR -e newuidmap shadow-4.1.5.1-r1/work/shadow-4.1.5.1/ finds nothing as well).

the same holds for newgidmap.c

```
--- shadow-4.1.5.1-r1/work/shadow-4.1.5.1/lib/commonio.c        2012-05-18 21:44:54.000000000 +0200
+++ shadow-4.2.1/work/shadow-4.2.1/lib/commonio.c       2014-05-08 10:43:11.000000000 +0200
@@ -32,7 +32,7 @@
 
 #include <config.h>
 
-#ident "$Id: commonio.c 3727 2012-05-18 19:44:53Z nekral-guest $"
+#ident "$Id$"
 
 #include "defines.h"
 #include <assert.h>
@@ -1113,6 +1113,38 @@
        return 1;
 }
 
+#ifdef ENABLE_SUBIDS
+int commonio_append (struct commonio_db *db, const void *eptr)
+{
+       struct commonio_entry *p;
+       void *nentry;
+
+       if (!db->isopen || db->readonly) {
+               errno = EINVAL;
+               return 0;
+       }
+       nentry = db->ops->dup (eptr);
+       if (NULL == nentry) {
+               errno = ENOMEM;
+               return 0;
+       }
+       /* new entry */
+       p = (struct commonio_entry *) malloc (sizeof *p);
+       if (NULL == p) {
+               db->ops->free (nentry);
+               errno = ENOMEM;
+               return 0;
+       }
+
+       p->eptr = nentry;
+       p->line = NULL;
+       p->changed = true;
+       add_one_entry (db, p);
+
+       db->changed = true;
+       return 1;
+}
+#endif                         /* ENABLE_SUBIDS */
 
 void commonio_del_entry (struct commonio_db *db, const struct commonio_entry *p)
 {
```

**Comment by hallyn** [comment]
Ok...  lxc falls back to writing uid_map entries by
hand if newuidmap does not exist.  So if it only
fails when newuidmap exists, then either newuidmap
in your shadow package is bad, or, more likely, you
do not have /etc/subuid and /etc/subgid entries for
the root user for the range used by the containers.

**Comment by specing** [comment]
The second option is not the case (checked by re-running usermod -v ... -w ... root on blank /etc/sub*), though on shadow 4.1.5.1-r1 it works even without these two files being present.

**Comment by hallyn** [comment]
Does grsecurity add a check that you must be the parent of a
task to set its uidmap?  Using newuidmap means that if 9021
clones 9022 in a new userns, it will then fork+exec 9023
to write to 9022's uidmap.

**Comment by specing** [comment]
I cannot answer that [yet], would require digging into the kernel. Meanwhile, 4.2.1-r2 was tested and it does not work either.

**Comment by jirutka** [comment]
@specing This is caused by grsecurity feature `GRKERNSEC_SYSFS_RESTRICT`, see https://github.com/lxc/lxc/issues/482#issuecomment-137612553.

**Comment by ragavaluvijay** [comment]
@specing and all,
I also faced similar issue  as shown below error when i try to create un-priv lxc container.

> lxc-create -t busybox -n Contname
> error mapping child
> setgid: Invalid argument
> lxc-create: lxccontainer.c: do_create_container_dir: 978 Failed to chown container dir
> lxc-create: lxc_create.c: main: 318 Error creating container Contname.

I am using kernel version 3.14 on busybox environment  and using LXC ver 2.0.4 . I noticed that it does not create or update file /etc/subuid & /etc/subgid with mapped uid ranges  when i create lxc user(which is low privilege user for un-privilege container creation ,basically it does not have sudo user permission). then i manually created these two files subuid/subgid file with below information
```
in subuid file , usrlxc:100000:65536 
in subgid file , usrlxc:100000:65536
```
and also noted that usermod does not have option to add/del subuid and subgid range values for the user, that i see usermod has only below options

> 
> Usage: usermod [options] LOGIN
> 
> Options:
>   -c, --comment COMMENT         new value of the GECOS field
>   -d, --home HOME_DIR           new home directory for the user account
>   -e, --expiredate EXPIRE_DATE  set account expiration date to EXPIRE_DATE
>   -f, --inactive INACTIVE       set password inactive after expiration
>                                 to INACTIVE
>   -g, --gid GROUP               force use GROUP as new primary group
>   -G, --groups GROUPS           new list of supplementary GROUPS
>   -a, --append                  append the user to the supplemental GROUPS
>                                 mentioned by the -G option without removing
>                                 him/her from other groups
>   -h, --help                    display this help message and exit
>   -l, --login NEW_LOGIN         new value of the login name
>   -L, --lock                    lock the user account
>   -m, --move-home               move contents of the home directory to the
>                                 new location (use only with -d)
>   -o, --non-unique              allow using duplicate (non-unique) UID
>   -p, --password PASSWORD       use encrypted password for the new password
>   -s, --shell SHELL             new login shell for the user account
>   -u, --uid UID                 new UID for the user account
>   -U, --unlock                  unlock the user account
>   -Z, --selinux-user            new SELinux user mapping for the user account 

also , i dont find "newuidmap" or " newuidmap"  utility in my environment . 
Can you please let me know if it something missing in my kernel or lxc config

**Comment by brauner** [comment]
`man shadow` and at the end of the document you usually find the version in left corner.

**Comment by ragavaluvijay** [comment]
@brauner , Thanks , 
I see below info in my ubuntu 16.04 machine . in where un-priv lxc is working fine with shadow 4.2 version
**shadow-utils 4.2**                  03/29/2016                         SHADOW(5)

do we have any other option to know the shadow version as i am using busybox environment with customized linux kernel inwhere "man" page support is not available.

**Comment by ragavaluvijay** [comment]
@brauner ,
I am re-phrasing my question  
Basically , I would like to know what might have went wrong in my environment as i found below three in my environment which cause error while creating un-priv container.

> No support newuidmap and newgidmap support. 
> "adduser" does not create/update subuid/subgid files while i create new user
>  "usermod" does not have option to change subuid/subgid ranges 

and  in my same environment , i dont have any issue with privilege container creation under root and sudo user . i could create/start/attach the container and execute required processes inside privilege container from root/sudo user
 only issue is i am facing is , for un-privilege container at my same environment .

Please provide your comment if you think any issue with my kernel or lxc config fro this issue

**Comment by brauner** [comment]
Afaik, we require the presence of `newuidmap` and `newgidmap` for unprivileged `LXC` to work correctly. So if you can you should update to at least `shadow` version `4.2.1`.

**Comment by ragavaluvijay** [comment]
@brauner,
after upgrading shadow version to 4.2.1-r0 from 4.1.4.3-r14 in my target platform , the above mentioned issues are resolved. i could able to create Un-privilege container . Thanks a lot :-)
I will start and execute some sample process in the container and let you know

**Comment by brauner** [comment]
Thanks for letting us know. Closing this issue then. :)

**Comment by ragavaluvijay** [comment]
please close this ,
unfortunately i am facing issue for lxc-start of un-privilege container .
 anyway this is not related to newuidmap or newgidmap .
i will post newly if i could resolve by myown. the issue related to cgroups

**Comment by specing** [comment]
@jirutka Caused? no. I will still have try disabling that config option, but this is not the way forward.
User namespace containers used to work with previous shadow packages on the same kernel, so the bug must be in shadow, right? Or if I word it in another way, shadow does not support grsecurity [anymore].

When attempting to start a container today I was again visited by my life-long friend newuidmap. This time however, I was not able to find an earlier version ebuild. Therefore, I took a more invasive approach:

    if ((getuid() != pw->pw_uid) ||
        (getgid() != pw->pw_gid) ||
        (pw->pw_uid != st.st_uid) ||
        (pw->pw_gid != st.st_gid)) {
        fprintf(stderr, _( "%s: Target process %u is owned by a different user: uid:%lu pw_uid:%lu st_uid:%lu, gid:%lu pw_gid:%lu st_gid:%lu\n" ),
            Prog, target,
            (unsigned long int)getuid(), (unsigned long int)pw->pw_uid, (unsigned long int)st.st_uid,
            (unsigned long int)getgid(), (unsigned long int)pw->pw_gid, (unsigned long int)st.st_gid);
        if (getuid() == 0 && pw->pw_uid == 0 && getgid() == 0 && pw->pw_gid == 0)
            fprintf(stderr, _( "%s: root can do everything\n" ), Prog);
        else
            return EXIT_FAILURE;
    }

The above should only work when the user is actually root. Any other suggestions that do not involve disabling grsecurity?

**Comment by hallyn** [comment]
At this point I'm afraid this is hard to follow.  If it used to work with an older shadow, can you bisect to see which upstream commit breaks it for you?

**Comment by specing** [comment]
Sure, but my guess is that it is the commit that introduced this check.

**Comment by specing** [comment]
Confirmed, 2cffa14105a8002695d43a609fef918126892013 from https://github.com/shadow-maint/shadow.git works, while 673c2a6f9aa6c69588f4c1be08589b8d3475a520 (next commit) does not. The later introduced these suid helpers.

Ebuild used for building & installing them (modified from https://gitweb.gentoo.org/repo/gentoo.git/diff/sys-apps/shadow/shadow-4.2.1-r2.ebuild?id=4d5d0eac6f3ae936d0bdcd291ef01a39bfb8fd03): 
[shadow-9999.ebuild.txt](https://github.com/lxc/lxc/files/870087/shadow-9999.ebuild.txt)

**Comment by hallyn** [comment]
Ok, so are newuidmap and newgidmap setuid-root?

If not, then if making them so fixes it, then it should be a trivial fix to the ebuild.

**Comment by specing** [comment]
/o\

Yes, they are not suid root.

stat /usr/bin/newgidmap 
  File: '/usr/bin/newgidmap'
  Size: 32896           Blocks: 72         IO Block: 4096   regular file
Device: 12h/18d Inode: 872006      Links: 1
Access: (0755/-rwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2017-03-25 18:31:15.000000000 +0100
Modify: 2017-03-25 18:31:15.000000000 +0100
Change: 2017-03-25 18:31:27.923751241 +0100

**Comment by hallyn** [comment]
Ok, that - or at least filecaps - is required.   Someone with privilege must delegate the subuids.

**Comment by specing** [comment]
I've just tried recompiling non-patched shadow and then making /usr/bin/newuidmap and newgidmap
4711 with no effect. I am running these containers as root, so that should not matter anyway.

**Comment by hallyn** [comment]
Meaning it failed the exact same way?

So does just doing
	sudo mv /usr/bin/newuidmap{,.1}
	sudo mv /usr/bin/newgidmap{,.1}
fix it?

My guess then is grsecurity refuses writing to another task's
/proc/$$/uid_map file regardless of privilege.  If you care
enough I'd recommend looking through the proc and trace
restrictions it adds.

**Comment by specing** [comment]
Yes, it failed the exact same way and yes, that "fixes" it.

**Comment by hallyn** [comment]
So it seems worthwhile in lxc to, on failure, check for filecaps or setuid-root on newuidmap and return a meaningful error.
